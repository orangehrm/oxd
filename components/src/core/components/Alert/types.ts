/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */

export const TYPE_DEFAULT = 'default';
export const TYPE_SUCCESS = 'success';
export const TYPE_WARN = 'warn';
export const TYPE_ERROR = 'error';
export const TYPE_INFO = 'info';

export const TYPES = [
  TYPE_DEFAULT,
  TYPE_SUCCESS,
  TYPE_WARN,
  TYPE_ERROR,
  TYPE_INFO,
];

export const ICON_MAP = {
  [TYPE_DEFAULT]: 'star',
  [TYPE_SUCCESS]: 'check2',
  [TYPE_WARN]: 'exclamation-triangle',
  [TYPE_ERROR]: 'exclamation-circle',
  [TYPE_INFO]: 'info-circle',
};

export type AlertType =
  | typeof TYPE_DEFAULT
  | typeof TYPE_SUCCESS
  | typeof TYPE_WARN
  | typeof TYPE_ERROR
  | typeof TYPE_INFO;
