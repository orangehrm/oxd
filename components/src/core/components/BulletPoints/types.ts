export interface BulletPointListItem {
    id: string;
    styles?: object;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    [key: string]: any;
  }