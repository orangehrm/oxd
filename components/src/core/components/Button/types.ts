export const SIZE_SMALL = 'small';
export const SIZE_MEDIUM = 'medium';
export const SIZE_LARGE = 'large';
export const SIZE_LONG = 'long';
export const SIZES = [SIZE_SMALL, SIZE_MEDIUM, SIZE_LARGE, SIZE_LONG];

export const TYPE_MAIN = 'main';
export const TYPE_SECONDARY = 'secondary';
export const TYPE_INFO = 'info';
export const TYPE_DANGER = 'danger';
export const TYPE_WARN = 'warn';
export const TYPE_SUCCESS = 'success';

export const TYPE_GHOST = 'ghost';
export const TYPE_GHOST_INFO = 'ghost-info';
export const TYPE_GHOST_DANGER = 'ghost-danger';
export const TYPE_GHOST_WARN = 'ghost-warn';
export const TYPE_GHOST_SUCCESS = 'ghost-success';

export const TYPE_LABEL = 'label';
export const TYPE_LABEL_INFO = 'label-info';
export const TYPE_LABEL_DANGER = 'label-danger';
export const TYPE_LABEL_WARN = 'label-warn';
export const TYPE_LABEL_SUCCESS = 'label-success';

export const TYPE_TEXT = 'text';
export const TYPE_TOOL = 'tool';

export const TYPE_GLASS = 'glass';

export const TOOLTIP_TOP = 'top';
export const TOOLTIP_BOTTOM = 'bottom';
export const TOOLTIP_LEFT = 'left';
export const TOOLTIP_RIGHT = 'right';

export const TOOLTIP_POSITIONS = [
  TOOLTIP_TOP,
  TOOLTIP_BOTTOM,
  TOOLTIP_LEFT,
  TOOLTIP_RIGHT,
];

export const TYPES = [
  TYPE_MAIN,
  TYPE_SECONDARY,
  TYPE_INFO,
  TYPE_DANGER,
  TYPE_WARN,
  TYPE_SUCCESS,
  TYPE_GHOST,
  TYPE_GHOST_INFO,
  TYPE_GHOST_DANGER,
  TYPE_GHOST_WARN,
  TYPE_GHOST_SUCCESS,
  TYPE_LABEL,
  TYPE_LABEL_INFO,
  TYPE_LABEL_DANGER,
  TYPE_LABEL_WARN,
  TYPE_LABEL_SUCCESS,
  TYPE_TEXT,
  TYPE_TOOL,
  TYPE_GLASS,
];

export type ButtonSize =
  | typeof SIZE_SMALL
  | typeof SIZE_MEDIUM
  | typeof SIZE_LARGE
  | typeof SIZE_LONG;

export type ButtonType =
  | typeof TYPE_MAIN
  | typeof TYPE_SECONDARY
  | typeof TYPE_INFO
  | typeof TYPE_DANGER
  | typeof TYPE_WARN
  | typeof TYPE_SUCCESS
  | typeof TYPE_GHOST
  | typeof TYPE_GHOST_INFO
  | typeof TYPE_GHOST_DANGER
  | typeof TYPE_GHOST_WARN
  | typeof TYPE_GHOST_SUCCESS
  | typeof TYPE_LABEL
  | typeof TYPE_LABEL_INFO
  | typeof TYPE_LABEL_DANGER
  | typeof TYPE_LABEL_WARN
  | typeof TYPE_LABEL_SUCCESS
  | typeof TYPE_TEXT
  | typeof TYPE_TOOL;

export type TooltipPosition =
  | typeof TOOLTIP_TOP
  | typeof TOOLTIP_BOTTOM
  | typeof TOOLTIP_LEFT
  | typeof TOOLTIP_RIGHT;
