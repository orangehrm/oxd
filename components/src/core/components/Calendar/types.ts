/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */

export const SUNDAY = 'S';
export const MONDAY = 'M';
export const TUESDAY = 'T';
export const WEDNESDAY = 'W';
export const THURSDAY = 'T';
export const FRIDAY = 'F';
export const SATURDAY = 'S';

export const JANUARY = 'January';
export const FEBRUARY = 'February';
export const MARCH = 'March';
export const APRIL = 'April';
export const MAY = 'May';
export const JUNE = 'June';
export const JULY = 'July';
export const AUGUST = 'August';
export const SEPTEMBER = 'September';
export const OCTOBER = 'October';
export const NOVEMBER = 'November';
export const DECEMBER = 'December';

export const DAYS_OF_WEEK = [
  SUNDAY,
  MONDAY,
  TUESDAY,
  WEDNESDAY,
  THURSDAY,
  FRIDAY,
  SATURDAY,
];

export const MONTHS_OF_YEAR = [
  JANUARY,
  FEBRUARY,
  MARCH,
  APRIL,
  MAY,
  JUNE,
  JULY,
  AUGUST,
  SEPTEMBER,
  OCTOBER,
  NOVEMBER,
  DECEMBER,
];

export type CalendarEvent = {
  date: Date;
  type: string;
  class?: string;
  style?: object;
};

export interface CalendarDayAttributes {
  index: number;
  class?: string;
  style?: object;
}
