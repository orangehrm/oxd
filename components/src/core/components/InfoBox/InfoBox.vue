<template>
  <div class="oxd-select-wrapper" :class="{'--selected': modelValue}">
    <button
      v-bind="$attrs"
      class="oxd-select-fill-container"
      :style="infoBoxContainerStyles"
      :class="classes"
      @focus="onFocus"
      @blur="onBlur"
      @keyup.esc="onCloseDropdown"
      @keydown.enter.prevent="onSelectEnter"
      @keydown.down.exact.prevent="onSelectDown"
      @keydown.up.exact.prevent="onSelectUp"
      @keydown="onKeypress"
      @click.prevent=""
      v-click-outside="clickOutside"
    >
      <div class="d-flex flex-wrap w-100 h-100">
        <div
          class="
            d-flex
            align-start
            justify-start
            w-100
            oxd-select-fill-title-row
            --mb
          "
          :style="titleWrapperStyles"
        >
          <label class="oxd-select-fill-title" :style="titleLabelStyles">{{
            $vt(infoLabel)
          }}</label>
        </div>
        <div
          class="
            d-flex
            align-end
            justify-between
            oxd-select-fill-subtitle-row
            w-100
          "
        >
          <div
            class="
              d-flex
              align-start
              justify-start
              oxd-select-fill-subtitle-wrapper
            "
            :style="subtitleWrapperStyles"
          >
            <label
              class="oxd-select-fill-subtitle text-left"
              :style="selectedItemLabelStyles"
              >{{ $vt(getLabel) }}</label
            >
          </div>
          <div
            v-if="!(isModelValueString || disabled)"
            class="d-flex align-center justify-start"
          >
            <oxd-icon-button
              :name="dropdownOpen ? 'oxd-chevron-up' : 'oxd-chevron-down'"
              size="xxx-small"
              class="oxd-select-dropdown-trigger"
              :class="{'cursor-default': readonly}"
              :style="infoBoxTriggerButtonStyles"
              :disabled="readonly"
              @click="onToggleDropdown"
            />
          </div>
        </div>
      </div>
    </button>
    <oxd-select-dropdown
      v-if="dropdownOpen"
      :class="dropdownClasses"
      :style="dropdownStyles"
      :loading="loading"
      :empty="computedOptions.length === 0"
    >
      <oxd-select-option
        v-for="(option, i) in computedOptions"
        :key="option.id"
        :class="optionClasses[i]"
        :disabled="option._disabled || option._selected"
        :ref="`option-${i}`"
        @select="onSelect(option)"
      >
        <slot name="option" :data="option"></slot>
        <span v-if="!$slots['option']">{{ $vt(option.label) }}</span>
      </oxd-select-option>
    </oxd-select-dropdown>
  </div>
</template>

<script lang="ts">
import {defineComponent} from 'vue';
import eventsMixin from './../Input/Select/events-mixin';
import navigationMixin from './../Input/Select/navigation-mixin';
import translateMixin from '../../../mixins/translate';
import {
  TOP,
  BOTTOM,
  LEFT,
  RIGHT,
  Option,
  DROPDOWN_POSITIONS,
} from './../Input/types';
import clickOutsideDirective from './../../../directives/click-outside';
import {TooltipPosition} from './../Button/types';
import SelectDropdown from '@orangehrm/oxd/core/components/Input/Select/SelectDropdown.vue';
import SelectOption from '@orangehrm/oxd/core/components/Input/Select/SelectOption.vue';
import ButtonIcon from '@orangehrm/oxd/core/components/Button/Icon.vue';
import {hexToRgb} from './../../../utils/colorConverter';

export default defineComponent({
  name: 'oxd-infobox',
  inheritAttrs: false,

  components: {
    'oxd-icon-button': ButtonIcon,
    'oxd-select-dropdown': SelectDropdown,
    'oxd-select-option': SelectOption,
  },

  mixins: [navigationMixin, eventsMixin, translateMixin],

  directives: {
    'click-outside': clickOutsideDirective,
  },

  emits: [
    'update:modelValue',
    'dropdown:opened',
    'dropdown:closed',
    'dropdown:blur',
    'dropdown:clear',
  ],

  props: {
    modelValue: {
      type: Object,
    },
    hasError: {
      type: Boolean,
      default: false,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    readonly: {
      type: Boolean,
      default: false,
    },
    options: {
      type: Array,
      default: () => [],
    },
    dropdownPosition: {
      type: String,
      default: BOTTOM,
      validator: function (value: TooltipPosition) {
        return DROPDOWN_POSITIONS.indexOf(value) !== -1;
      },
    },
    dropdownStyles: {
      type: Object,
      default: () => null,
    },
    numOfTitleRows: {
      type: Number,
      default: 1,
    },
    numOfValueRows: {
      type: Number,
      default: 2,
    },
    infoLabel: {
      type: String,
    },
    titleLineHeight: {
      type: Number,
      default: () => 22,
    },
    subtitleLineHeight: {
      type: Number,
      default: () => 19,
    },
  },

  data() {
    return {
      focused: false,
      loading: false,
      dropdownOpen: false,
    };
  },

  computed: {
    computedOptions(): Option[] {
      return this.options.map((option: Option) => {
        let _selected = false;
        if (this.modelValue?.id === option.id) {
          _selected = true;
        }
        return {...option, _selected};
      });
    },
    dropdownClasses(): object {
      return {
        '--positon-bottom': this.dropdownPosition === BOTTOM,
        '--positon-top': this.dropdownPosition === TOP,
        '--positon-left': this.dropdownPosition === LEFT,
        '--positon-right': this.dropdownPosition === RIGHT,
      };
    },
    optionClasses(): object[] {
      return this.computedOptions.map((option: Option, index: number) => {
        return {
          '--disabled': option._disabled,
          '--selected': option._selected,
          '--focused': index === this.pointer,
          [`--indent-${option._indent}`]: option._indent !== undefined,
        };
      });
    },
    titleLabelStyles(): {
      '-webkit-line-clamp'?: number;
    } {
      return {
        '-webkit-line-clamp': this.numOfTitleRows,
      };
    },
    selectedItemLabelStyles(): {
      color: string | null;
      'font-weight'?: number;
      '-webkit-line-clamp'?: number;
    } {
      return {
        color: hexToRgb(this.modelValue?.color),
        '-webkit-line-clamp': this.numOfValueRows,
        'font-weight': this.modelValue?.color ? 700 : 600,
      };
    },
    infoBoxContainerStyles(): {
      'background-color': string | null;
    } {
      return {
        'background-color': hexToRgb(this.modelValue?.color, '7%'),
      };
    },
    infoBoxTriggerButtonStyles(): {
      color: string | null;
      'background-color': string | null;
    } {
      return {
        color: hexToRgb(this.modelValue?.color),
        'background-color': hexToRgb(this.modelValue?.color, '10%'),
      };
    },
    isModelValueString(): boolean {
      return typeof this.modelValue === 'string';
    },
    // modelValue can be anything
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    getLabel(): any {
      if (this.isModelValueString) {
        return this.modelValue;
      } else {
        return this.modelValue ? this.modelValue.label : null;
      }
    },
    subtitleWrapperStyles(): {
      height: string;
    } {
      const subtitleWrapperHeight =
        this.numOfValueRows * this.subtitleLineHeight;
      return {
        height: `${subtitleWrapperHeight}px`,
      };
    },
    titleWrapperStyles(): {
      height: string;
    } {
      const titleWrapperHeight = this.numOfTitleRows * this.titleLineHeight;
      return {
        height: `${titleWrapperHeight}px`,
      };
    },
    classes(): object {
      return {
        'oxd-select-fill-container--focus': this.focused,
        'oxd-select-fill-container--error': this.hasError,
        'oxd-select-fill-container--disabled': this.disabled,
        'oxd-select-fill-container--readonly': this.readonly,
      };
    },
  },

  methods: {
    onFocus($e: Event) {
      if (this.disabled) {
        $e.stopImmediatePropagation();
      } else {
        this.focused = true;
      }
    },
    onBlur() {
      this.focused = false;
    },
    clickOutside() {
      this.dropdownOpen = false;
    },
  },
});
</script>

<style src="./info-box.scss" lang="scss" scoped></style>
