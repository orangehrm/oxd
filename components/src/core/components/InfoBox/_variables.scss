$oxd-select-input-button-color--secondary-pressed: #68a61d !default;
$oxd-info-box-title-color: $oxd-interface-gray-darken-1-color !default;
$oxd-info-box-font-family: $oxd-font-family !default;
$oxd-info-box-title-font-size: 16px !default;
$oxd-info-box-sub-title-font-size: 14px !default;
$oxd-info-box-font-weight: 700 !default;