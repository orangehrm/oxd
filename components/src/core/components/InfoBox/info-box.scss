@import '../../../styles';
@import 'variables';
@import './../Input/variables';

$dropdown-button-size: 24px;

.oxd-select-wrapper {
  @include oxd-input-control-size;
  position: relative;
  &:focus {
    outline: none;
  }
  & > *:focus {
    outline: none;
  }
  .oxd-select-dropdown-trigger {
    min-width: $dropdown-button-size;
    min-height: $dropdown-button-size;
    &.dropdown-minimized {
      margin-left: 0.5rem;
    }
    :deep(.oxd-svg-icon) {
      margin-right: 0;
    }
  }
}

.oxd-select-fill-container {
  box-sizing: border-box;
  position: relative;
  border: none;
  background: rgba(100, 114, 140, 7%);
  width: 100%;
  padding: 10px;
  border-radius: 0.85rem;
  cursor: auto !important;

  &--focus {
    border: $oxd-input-border--focus;
    box-shadow: $oxd-input-box-shadow--focus;
    outline: 0;
  }

  &--error {
    border: $oxd-input-border--error;
    box-shadow: $oxd-input-box-shadow--error;
  }

  &--disabled,
  &--readonly {
    background: $oxd-dropdown-dropdown-background--disabled;
  }

  .oxd-select-fill-title-row {
    &.--mb {
      margin-bottom: 0.25rem;
    }
    .oxd-select-fill-title {
      display: -webkit-box;
      -webkit-box-orient: vertical;
      overflow: hidden;
    }
  }
  .oxd-select-fill-subtitle-wrapper {
    box-sizing: border-box;
    width: calc(100% - 24px);
    padding-right: 5px;
    .oxd-select-fill-subtitle {
      display: -webkit-box;
      -webkit-box-orient: vertical;
      overflow: hidden;
      overflow-wrap: break-word;
      margin-bottom: 0;
    }
    .oxd-info-box-clickable-text {
      font-size: 14px;
      font-weight: 600;
      color: #64728c;
      text-decoration: underline;
      font-style: italic;
      cursor: pointer;
      text-align: left;
      display: -webkit-box;
      -webkit-box-orient: vertical;
      overflow: hidden;
      overflow-wrap: break-word;
      -webkit-line-clamp: 1;
      width: 100%;
    }
  }
  .oxd-select-fill {
    &-title,
    &-subtitle {
      font-family: $oxd-info-box-font-family;
      color: $oxd-info-box-title-color;
      font-size: $oxd-info-box-title-font-size;
      font-weight: $oxd-info-box-font-weight;
    }
    &-title {
      text-align: left;
    }
    &-subtitle {
      font-weight: $oxd-info-box-font-weight;
      font-size: $oxd-info-box-sub-title-font-size;
    }
  }
}

.oxd-select-dropdown {
  position: absolute;
  width: inherit;
  box-sizing: border-box;
  color: $oxd-dropdown-dropdown-color;
  z-index: $oxd-dropdown-dropdown-z-index;
  border: $oxd-dropdown-dropdown-border;
  padding: $oxd-dropdown-dropdown-padding;
  box-shadow: $oxd-dropdown-dropdown-box-shadow;
  border-radius: $oxd-dropdown-dropdown-border-radius;
  background-color: $oxd-dropdown-dropdown-background;

  &.--positon-top {
    bottom: 100%;
    margin-bottom: $oxd-dropdown-dropdown-margin;
  }

  &.--positon-right {
    left: 100%;
    margin-left: $oxd-dropdown-dropdown-margin;
  }

  &.--positon-left {
    right: 100%;
    margin-right: $oxd-dropdown-dropdown-margin;
  }

  &.--positon-bottom {
    margin-top: $oxd-dropdown-dropdown-margin;
  }

  .oxd-select-dropdown-inner {
    overflow-y: auto;
    overflow-x: hidden;
    padding: $oxd-dropdown-dropdown-inner-padding;
    max-height: $oxd-dropdown-dropdown-max-height;
    @include oxd-scrollbar();
  }

  &.--with-empty-selector .oxd-select-dropdown-inner {
    max-height: $oxd-dropdown-dropdown-with-empty-selector-max-height;
  }
}

.oxd-select-option {
  border: 0;
  display: flex;
  align-items: center;
  text-align: inherit;
  cursor: pointer;
  color: $oxd-dropdown-option-color;
  font-size: $oxd-dropdown-option-font-size;
  font-family: $oxd-dropdown-option-font-family;
  font-weight: $oxd-dropdown-option-font-weight;
  padding: $oxd-dropdown-option-padding;
  margin: $oxd-dropdown-option-margin;
  background-color: $oxd-dropdown-option-background;
  min-height: $oxd-dropdown-option-min-height;

  span {
    word-break: break-word;
    display: -webkit-box;
    -webkit-line-clamp: 3;
    -webkit-box-orient: vertical;
    overflow: hidden;
    text-overflow: ellipsis;
  }

  &:last-child {
    border-bottom: unset;
  }
  &:hover,
  &.--focused {
    border-radius: $oxd-dropdown-option-border-radius;
    background-color: $oxd-dropdown-option-background--focused;
  }
  &.--disabled {
    border-radius: $oxd-dropdown-option-border-radius;
    color: $oxd-dropdown-option-color--disabled;
    cursor: unset;
    &:hover,
    &.--focused {
      background-color: transparent;
    }
  }
  &.--selected {
    border-radius: $oxd-dropdown-option-border-radius;
    background-color: $oxd-dropdown-option-background--focused;
    cursor: unset;
  }
  @for $i from 1 through 20 {
    $value: $i * 0.5;
    &.--indent-#{$i} {
      padding-left: #{$value}rem;
    }
  }
}
.no-label {
  min-width: initial;
  justify-content: center !important;
  ::v-deep(.oxd-button-label-wrapper) {
    display: none;
  }
  ::v-deep(.oxd-icon.bi-chevron-down) {
    display: none !important;
  }
  ::v-deep(.oxd-icon.bi-chevron-up) {
    display: none !important;
  }
}
.dropdown-btn {
  display: flex;
  align-items: center;
  justify-content: space-between;
  font-size: 12px;
  border-radius: 0.85rem;
  &.button-double-line {
    padding: 0.5rem 0.75rem 0.5rem 1rem;
    cursor: default !important;
    :deep(.oxd-button-label-wrapper) {
      overflow: initial;
      white-space: initial;
      text-overflow: initial;
      width: 100%;
      font-weight: 800;
      .label-double-line-wrapper.label-double-line .label {
        overflow: initial;
        text-overflow: initial;
        white-space: initial;
        word-break: break-word;
        text-align: left;
        padding-right: 5px;
      }
    }
  }
}
.oxd-icon {
  &.oxd-button-icon-right {
    margin-left: 0.5rem;
  }
}
.mini-icon-left {
  margin-right: 0.5rem;
  color: $oxd-select-input-button-color--secondary-pressed;
}
.oxd-select-info-button-container {
  min-width: $dropdown-button-size;
  .oxd-select-info-button {
    margin-right: 0 !important;
    cursor: pointer;
    :deep(path) {
      fill: $oxd-select-input-button-color--secondary-pressed;
    }
  }
}
.label-double-line-wrapper {
  vertical-align: middle;
  display: inline-grid;
  &.label-double-line {
    width: 100%;
    .label-small {
      margin-bottom: 0.25rem;
      font-size: 0.625rem;
      line-height: 1;
      font-weight: 600;
    }
    .label {
      line-height: 0.75rem;
      font-size: 0.75rem;
      font-weight: 800;
      max-width: calc(100% - 32px);
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
      display: inline-block;
    }
  }
}
