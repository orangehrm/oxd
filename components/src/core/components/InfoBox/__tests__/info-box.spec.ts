import { mount, shallowMount } from '@vue/test-utils';
import InfoBox from '@orangehrm/oxd/core/components/InfoBox/InfoBox.vue';
import ButtonIcon from '@orangehrm/oxd/core/components/Button/Icon.vue';
import SelectOption from '@orangehrm/oxd/core/components/Input/Select/SelectOption.vue';
import { BOTTOM } from '../../Input/types';

const options = [
  {
    id: 2,
    label: 'Application Received',
  },
  {
    id: 3,
    label: 'Shortlisted',
  },
  {
    id: 4,
    label: 'In Progress',
  },
  {
    id: 5,
    label: 'Job Offer',
  },
  {
    id: 6,
    label: 'Hired',
    color: '#68a61d',
  },
  {
    id: 7,
    label: 'Rejected',
    color: '#eb0910',
  },
  {
    id: 7,
    label: 'orem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa quis enim. Donec pede justo, fringilla vel, aliquet nec, vulputate eget, arcu. In enim justo, rhoncus ut, imperdiet a, venenatis vitae, justo. Nullam dictum felis eu pede mollis pretium. Integer tincidunt. Cras dapibus.'
  }
];
const infoLabel = 'Current Stage of Recruitment';
const numOfTitleRows = 2;
const numOfValueRows = 3;
const dropdownPosition = BOTTOM;

describe('InfoBox.vue', () => {

  it('renders OXD Select Input', () => {
    const wrapper = mount(InfoBox, {
      props: { infoLabel, numOfTitleRows, numOfValueRows, options, dropdownPosition },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should load options to Select', async () => {
    const wrapper = mount(InfoBox, {
      props: { infoLabel, numOfTitleRows, numOfValueRows, options, dropdownPosition },
    });
    wrapper.findComponent(ButtonIcon).trigger('click');
    await wrapper.vm.$nextTick();
    const nodes = wrapper.findAllComponents(SelectOption);
    expect(nodes.length).toBe(7);
  });

  it('should select one option', async () => {
    const wrapper = mount(InfoBox, {
      props: { infoLabel, numOfTitleRows, numOfValueRows, options, dropdownPosition },
    });
    wrapper.findComponent(ButtonIcon).trigger('click');
    await wrapper.vm.$nextTick();
    const nodes = wrapper.findAllComponents(SelectOption);
    await nodes[1].trigger('mousedown');
    expect(wrapper.emitted('update:modelValue')).toEqual([
      [
        {
          id: 3,
          label: 'Shortlisted',
          _selected: false,
        },
      ],
    ]);
  });

  it('Dropdown trigger button should NOT be available when disabled', async () => {
    const wrapper = mount(InfoBox, {
      props: {
        infoLabel,
        numOfTitleRows,
        numOfValueRows,
        options,
        dropdownPosition,
        disabled: true,
      },
    });
    expect(wrapper.findComponent(ButtonIcon).exists()).toBeFalsy();
  });

  it('should close the dropdown when clicked outside', async () => {
    const wrapper = mount(InfoBox, {
      props: {
        infoLabel,
        numOfTitleRows,
        numOfValueRows,
        options,
        dropdownPosition,
      },
    });
    wrapper.vm.clickOutside();
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('dropdown:opened')).toBeFalsy();
    expect(wrapper.vm.dropdownOpen).toEqual(false);
    expect(wrapper.find('.oxd-select-dropdown').exists()).toBeFalsy();
  });

  it('should select one option with color attribute and return converted rgba color', async () => {
    const wrapper: any = mount(InfoBox, {
      props: { infoLabel, numOfTitleRows, numOfValueRows, options, dropdownPosition },
    });
    wrapper.setProps({
      modelValue: options[4]
    })
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.infoBoxContainerStyles).toStrictEqual({ 'background-color': 'rgba(104, 166, 29, 7%)' });
  });

  it('should select one option with color attribute and return converted rgba color', async () => {
    const wrapper: any = mount(InfoBox, {
      props: { infoLabel, numOfTitleRows, numOfValueRows, options, dropdownPosition },
    });
    wrapper.setProps({
      modelValue: '2022-05-03'
    })
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.getLabel).toStrictEqual('2022-05-03');
  });

  it('should set empty array as default when didnot pass options', async () => {
    const wrapper: any = shallowMount(InfoBox, {
      props: { infoLabel, numOfTitleRows, numOfValueRows, dropdownPosition },
    });
    wrapper.findComponent(ButtonIcon).trigger('click');
    await wrapper.vm.$nextTick();
    const nodes = wrapper.findAllComponents(SelectOption);
    expect(nodes.length).toBe(0);
  });

  it('should not open the dropdown when the button icon is clicked but the component is readonly', async () => {
    const wrapper = mount(InfoBox, {
      props: {
        infoLabel,
        numOfTitleRows,
        numOfValueRows,
        options,
        dropdownPosition,
        readonly: true,
      },
    });
    wrapper.findComponent(ButtonIcon).trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('dropdown:opened')).toBeFalsy();
    expect(wrapper.vm.dropdownOpen).toEqual(false);
    expect(wrapper.find('.oxd-select-dropdown').exists()).toBeFalsy();
  });

  it('Selected option should marked as selected true', async () => {
    const wrapper = mount(InfoBox, {
      props: { infoLabel, numOfTitleRows, numOfValueRows, options, dropdownPosition },
    });
    wrapper.setProps({
      modelValue: options[1]
    });
    wrapper.findComponent(ButtonIcon).trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.computedOptions[1]._selected).toStrictEqual(true);
  });

  it('on Focus it should NOT add a class "--focus" to the .oxd-select-fill-container when disabled', async () => {
    const wrapper = mount(InfoBox, {
      props: { infoLabel, numOfTitleRows, numOfValueRows, options, dropdownPosition, disabled: true },
    });
    const fillContainer = wrapper.find('.oxd-select-fill-container');
    fillContainer.trigger('focus');
    await wrapper.vm.$nextTick();
    expect(fillContainer.find('.oxd-select-fill-container--focus').exists()).toBe(
      false
    );
  });

  it('on Focus it should add a class "--readonly" to the .oxd-select-fill-container when readonly', async () => {
    const wrapper = mount(InfoBox, {
      props: { infoLabel, numOfTitleRows, numOfValueRows, options, dropdownPosition, readonly: true },
    });
    const fillContainer = wrapper.find('.oxd-select-fill-container');
    fillContainer.trigger('focus');
    await wrapper.vm.$nextTick();
    expect(fillContainer.find('.oxd-select-fill-container--readonly').exists()).toBe(
      true
    );
  });

  it('on Focus it should add a class "--focus" to the .oxd-select-fill-container', async () => {
    const wrapper = mount(InfoBox, {
      props: { infoLabel, numOfTitleRows, numOfValueRows, options, dropdownPosition },
    });
    const fillContainer = wrapper.find('.oxd-select-fill-container');
    fillContainer.trigger('focus');
    await wrapper.vm.$nextTick();
    expect(fillContainer.find('.oxd-select-fill-container--focus').exists()).toBe(
      true
    );
  });

  it('on Blur it should remove the class "--focus" from .oxd-select-fill-container', async () => {
    const wrapper = mount(InfoBox, {
      props: { infoLabel, numOfTitleRows, numOfValueRows, options, dropdownPosition },
    });
    wrapper.vm.onBlur();
    await wrapper.vm.$nextTick();
    expect(wrapper.find('.oxd-select-fill-container--focus').exists()).toBe(
      false
    );
  });

});
