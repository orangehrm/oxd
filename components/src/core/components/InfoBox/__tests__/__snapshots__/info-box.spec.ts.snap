// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfoBox.vue renders OXD Select Input 1`] = `
<div class="oxd-select-wrapper"><button class="oxd-select-fill-container">
    <div class="d-flex flex-wrap w-100 h-100">
      <div class="d-flex align-start justify-start w-100 oxd-select-fill-title-row --mb" style="height: 44px;"><label class="oxd-select-fill-title">Current Stage of Recruitment</label></div>
      <div class="d-flex align-end justify-between oxd-select-fill-subtitle-row w-100">
        <div class="d-flex align-start justify-start oxd-select-fill-subtitle-wrapper" style="height: 57px;"><label class="oxd-select-fill-subtitle text-left" style="font-weight: 600;"></label></div>
        <div class="d-flex align-center justify-start"><button type="button" class="oxd-icon-button oxd-select-dropdown-trigger" flow="top"><i class="oxd-svg-icon oxd-svg-icon--xxx-small"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 209.45 129.51">
                <defs>
                  <style>
                    .cls-1 {
                      stroke: currentColor;
                      stroke-miterlimit: 10;
                      stroke-width: 9px;
                    }
                  </style>
                </defs>
                <g id="Layer_2" data-name="Layer 2">
                  <g id="Layer_1-2" data-name="Layer 1">
                    <g id="Layer_2-2" data-name="Layer 2">
                      <g id="Layer_1-2-2" data-name="Layer 1-2">
                        <path class="cls-1" d="M105,91.28c12.12-13.2,24.1-26.2,36-39.2l38.23-41.64c3.49-3.81,7.78-5.83,13.2-5.18a13.48,13.48,0,0,1,11.48,8.41c2.08,5,1,9.59-2.68,13.64-6.12,6.73-12.3,13.42-18.45,20.12Q149.89,83.23,117,119.11c-4.07,4.45-8.85,6.83-15.24,5.55a14.67,14.67,0,0,1-8-4.69L8,26.5A12.52,12.52,0,0,1,9.27,8l.21-.18c6.25-5,14.9-4.24,20.47,1.83Q63.1,45.68,96.18,81.77C99.06,84.9,102,88,105,91.28Z"></path>
                      </g>
                    </g>
                  </g>
                </g>
              </svg></i></button></div>
      </div>
    </div>
  </button>
  <!--v-if-->
</div>
`;
